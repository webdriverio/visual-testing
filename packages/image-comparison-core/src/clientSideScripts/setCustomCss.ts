import type { CssOptions } from './customCss.interfaces.js'

/**
 * Set some default css
 */
export default function setCustomCss(cssOptions: CssOptions): void {
    if (!document.head) {
        return
    }
    // disabling CSS animations for everything including pseudo elements
    const disableTransformationsTransitionsAnimations = `
*, *::before, *::after {
    -o-transition-property: none !important;
    -moz-transition-property: none !important;
    -ms-transition-property: none !important;
    -webkit-transition-property: none !important;
    transition-property: none !important;
    -webkit-animation: none !important;
    -moz-animation: none !important;
    -o-animation: none !important;
    -ms-animation: none !important;
    animation: none !important;
    caret-color: transparent !important;
}`
    const { addressBarPadding, disableBlinkingCursor, disableCSSAnimation, id, toolBarPadding } = cssOptions
    const bodyTopPadding = addressBarPadding === 0 ? '' : `body{padding-top:${addressBarPadding}px !important}`
    const bodyBottomPadding = toolBarPadding === 0 ? '' : `body{padding-bottom:${toolBarPadding}px !important}`
    const disableBlinkingCursorCss = disableBlinkingCursor ? ' input, textarea, [contenteditable]{caret-color: transparent !important;}' : ''
    const css = (disableCSSAnimation ? disableTransformationsTransitionsAnimations : '') +
        bodyTopPadding +
        bodyBottomPadding +
        disableBlinkingCursorCss
    const head = document.head
    const style = document.createElement('style')

    style.id = id
    style.appendChild(document.createTextNode(css))
    head.appendChild(style)
}
