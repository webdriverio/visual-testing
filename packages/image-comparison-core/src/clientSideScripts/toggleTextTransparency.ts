/**
 * Add or remove the transparency from the text
 */
export default function toggleTextTransparency(enableTransparency: boolean): void {
    if (enableTransparency) {
        const elements = document.querySelectorAll<HTMLElement>('body *:not(script):not(style)')
        elements.forEach(el => {
            el.style.setProperty('color', 'transparent', 'important')
        })
        return
    }

    const elements = document.querySelectorAll<HTMLElement>('body [color]:not(script):not(style)')
    elements.forEach(el => {
        el.style.removeProperty('color')
    })
}
