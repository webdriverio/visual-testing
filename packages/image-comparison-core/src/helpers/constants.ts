import type { IosOffsets } from './constants.interfaces.js'
import type { ResizeDimensions } from '../methods/images.interfaces.js'
import type { TestContext } from 'src/methods/compareReport.interfaces.js'
import type { DeviceRectangles } from '../methods/rectangles.interfaces.js'

export const DEFAULT_COMPARE_OPTIONS = {
    blockOutSideBar: true,
    blockOutStatusBar: true,
    blockOutToolBar: true,
    createJsonReportFiles: false,
    diffPixelBoundingBoxProximity: 5,
    ignoreAlpha: false,
    ignoreAntialiasing: false,
    ignoreColors: false,
    ignoreLess: false,
    ignoreNothing: false,
    rawMisMatchPercentage: false,
    returnAllCompareData: false,
    saveAboveTolerance: 0,
    scaleImagesToSameSize: false,
}
export const DEFAULT_FORMAT_STRING = '{tag}-{browserName}-{width}x{height}-dpr-{dpr}'
export const STORYBOOK_FORMAT_STRING = '{tag}-{logName}-{width}x{height}-dpr-{dpr}'
export const FOLDERS = {
    ACTUAL: 'actual',
    DIFF: 'diff',
    TEMP_FULL_SCREEN: 'tempFullScreen',
    DEFAULT: {
        BASE: './__snapshots__/',
        SCREENSHOTS: '.tmp/',
    },
} as const
export const DEFAULT_SHADOW = {
    ADDRESS_BAR: 6,
    TOOL_BAR: 6,
}
export const DESKTOP = 'desktop'
export const NOT_KNOWN = 'not_known'
export const CUSTOM_CSS_ID = 'pic-css'
export const DEFAULT_RESIZE_DIMENSIONS: Required<ResizeDimensions> = {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
}
export const FULL_PAGE_SCROLL_TIMEOUT = 1500
export const DEFAULT_TABBABLE_OPTIONS = {
    circle: {
        backgroundColor: '#ff0000',
        borderColor: '#000',
        borderWidth: 1,
        fontColor: '#fff',
        fontFamily: 'Arial',
        fontSize: 10,
        size: 10,
        showNumber: true,
    },
    line: {
        color: '#000',
        width: 1,
    },
}
export const DEVICE_RECTANGLES: DeviceRectangles = {
    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
    homeBar: { y: 0, x: 0, width: 0, height: 0 },
    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
    statusBar: { y: 0, x: 0, width: 0, height: 0 },
    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
    screenSize: { width: 0, height: 0 },
    viewport: { y: 0, x: 0, width: 0, height: 0 },
}

/**
 * Android and iOS offsets data
 */
const ANDROID_DEFAULT_OFFSETS = {
    STATUS_BAR: 24,
    ADDRESS_BAR: 56,
    TOOL_BAR: 48,
}
export const ANDROID_OFFSETS = {
    6: { ...ANDROID_DEFAULT_OFFSETS },
    7: { ...ANDROID_DEFAULT_OFFSETS },
    8: { ...ANDROID_DEFAULT_OFFSETS },
    9: { ...ANDROID_DEFAULT_OFFSETS },
    10: { ...ANDROID_DEFAULT_OFFSETS },
    11: { ...ANDROID_DEFAULT_OFFSETS },
    12: { ...ANDROID_DEFAULT_OFFSETS },
    13: { ...ANDROID_DEFAULT_OFFSETS },
    14: { ...ANDROID_DEFAULT_OFFSETS },
}

//=========================
// iPhone/iPad screen data
//=========================
// Height of screen in portrait mode, this is also what you get back
// when you:
//  - ask for the screen dimensions in WebdriverIO `getWindowSize`
//  - with `screen.height` in Safari devtools
//  - with http://whatismyscreenresolution.net/
// You can also check:
// - mydevice.io
// - https://useyourloaf.com/archives/
// - https://blisk.io/devices
// Corners come from:
// - https://github.com/kylebshr/ScreenCorners
//   One point is the equivalent of 1.333(3) pixels.
//   On the other hand, one pixel is the equivalent of 0.75 points.
//------------------------------------------------------------------
// iPhones
// 6/7/8/SE/2020:                                667
// 6/7/8/Plus:                                   736
// X|XS|11Pro|12Mini|13Mini (notch):             812
// 12|12Pro|13|13Pro|14 (notch):                 844
// 14Pro (notch):                                852
// XSMax|XR|11|11ProMax (notch):                 896
// 12ProMax|13ProMax|14Plus (notch):             926
// 14ProMax (notch):                             932
//------------------------------------------------------------------
// iPads
// 9.7|Air 1/2|Mini 1/2/3/4/2019|Pro 9.7:       1024
// 10.2 1/2019/2020/2021                        1080
// Pro 10.5/2017/2019:                          1112
// Mini 2021 (notch):                           1133
// Air 2020/2022 (notch):                       1180
// Pro 11 2018/2021 (notch):                    1194
// Pro 12.9 2017 (no notch)/2018/2021 (notch):  1366
//------------------------------------------------------------------
export const IOS_OFFSETS: IosOffsets = {
    IPHONE: {
        // 6/7/8/SE/2020
        667: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        // 6/7/8/Plus
        736: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        // X|XS|11Pro|12Mini|13Mini (notch)
        812: {
            LANDSCAPE: {
                SAFE_AREA: 44,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 300, y: 359, height: 9, width: 212 },
            },
            PORTRAIT: {
                SAFE_AREA: 44,
                STATUS_BAR: 44,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 118, y: 796, height: 9, width: 137 },
            },
        },
        // 12|12Pro|13|13Pro|14 (notch)
        844: {
            LANDSCAPE: {
                SAFE_AREA: 47,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 312, y: 375, height: 9, width: 220 },
            },
            PORTRAIT: {
                SAFE_AREA: 47,
                STATUS_BAR: 47,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 124, y: 829, height: 9, width: 143 },
            },
        },
        // 14Pro|15|16 (dynamic island)
        852: {
            LANDSCAPE: {
                SAFE_AREA: 59,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 316, y: 378, height: 9, width: 220 },
            },
            PORTRAIT: {
                SAFE_AREA: 59,
                STATUS_BAR: 59,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 125, y: 837, height: 9, width: 143 },
            },
        },
        // 16 Pro (dynamic island)
        874: {
            LANDSCAPE: {
                SAFE_AREA: 50,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 324, y: 389, height: 9, width: 226 },
            },
            PORTRAIT: {
                SAFE_AREA: 62,
                STATUS_BAR: 62,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 129, y: 861, height: 9, width: 152 },
            },
        },
        // XSMax|XR|11|11ProMax (notch)
        896: {
            LANDSCAPE: {
                SAFE_AREA: 48,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 331, y: 399, height: 9, width: 234 },
            },
            PORTRAIT: {
                SAFE_AREA: 48,
                STATUS_BAR: 48,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 131, y: 881, height: 9, width: 152 },
            },
        },
        // 12ProMax|13ProMax|14Plus (notch)
        926: {
            LANDSCAPE: {
                SAFE_AREA: 47,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 342, y: 413, height: 9, width: 242 },
            },
            PORTRAIT: {
                SAFE_AREA: 47,
                STATUS_BAR: 47,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 136, y: 911, height: 9, width: 156 },
            },
        },
        // 14ProMax|16 Plus (notch)
        932: {
            LANDSCAPE: {
                SAFE_AREA: 59,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 345, y: 415, height: 9, width: 242 },
            },
            PORTRAIT: {
                SAFE_AREA: 59,
                STATUS_BAR: 59,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 137, y: 917, height: 9, width: 158 },
            },
        },
        // 16 Pro Max (dynamic island)
        956: {
            LANDSCAPE: {
                SAFE_AREA: 50,
                STATUS_BAR: 0,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 355, y: 427, height: 9, width: 246 },
            },
            PORTRAIT: {
                SAFE_AREA: 62,
                STATUS_BAR: 62,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 142, y: 943, height: 9, width: 158 },
            },
        },
    },
    IPAD: {
        // 9.7|Air 1/2|Mini 1/2/3/4/2019|Pro 9.7
        1024: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        // 10.2 1/2019/2020/2021
        1080: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        // Pro 10.5/2017/2019
        1112: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 20,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 0, y: 0, height: 0, width: 0 },
            },
        },
        // With Home bar
        // Mini 2021
        1133: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 408, y: 729, height: 9, width: 318 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 234, y: 1118, height: 9, width: 276 },
            },
        },
        // Air 2020/2022
        1180: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 431, y: 805, height: 9, width: 318 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 272, y: 1165, height: 9, width: 276 },
            },
        },
        // Pro 11 2018/2021
        1194: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 438, y: 819, height: 9, width: 318 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 279, y: 1179, height: 9, width: 276 },
            },
        },
        // Pro 12.9 2017/2018/2021
        1366: {
            LANDSCAPE: {
                SAFE_AREA: 0,
                // The iPad Pro (12.9 inch) (1st generation) has a status bar of 20px
                // STATUS_BAR: 24,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 508, y: 1009, height: 9, width: 350 },
            },
            PORTRAIT: {
                SAFE_AREA: 0,
                STATUS_BAR: 24,
                ADDRESS_BAR: 50,
                HOME_BAR: { x: 352, y: 1351, height: 9, width: 320 },
            },
        },
    },
}

export const supportedIosBezelDevices = [
    // iPhones
    'iphonex',
    'iphonexs',
    'iphonexsmax',
    'iphonexr',
    'iphone11',
    'iphone11pro',
    'iphone11promax',
    'iphone12',
    'iphone12mini',
    'iphone12pro',
    'iphone12promax',
    'iphone13',
    'iphone13mini',
    'iphone13pro',
    'iphone13promax',
    'iphone14',
    'iphone14plus',
    'iphone14pro',
    'iphone14promax',
    'iphone15',
    // // iPads
    'ipadmini', // 1133
    'ipadair', // 1180
    'ipadpro11', // 1194
    'ipadpro129', // 1366
]

export const DEFAULT_TEST_CONTEXT: TestContext = {
    commandName: 'Could not be determined',
    instanceData: {
        app: 'Could not be determined',
        browser: {
            name: 'Could not be determined',
            version: 'Could not be determined',
        },
        deviceName: 'Could not be determined',
        isMobile: false,
        isAndroid: false,
        isIOS: false,
        platform: {
            name: 'Could not be determined',
            version: 'Could not be determined',
        },
    },
    framework: 'mocha',
    parent: 'Could not be determined',
    tag: 'Could not be determined',
    title: 'Could not be determined',
}
