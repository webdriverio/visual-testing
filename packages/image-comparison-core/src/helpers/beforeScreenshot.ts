import logger from '@wdio/logger'
import hideScrollBars from '../clientSideScripts/hideScrollbars.js'
import setCustomCss from '../clientSideScripts/setCustomCss.js'
import { CUSTOM_CSS_ID } from './constants.js'
import { getAddressBarShadowPadding, getToolBarShadowPadding, waitFor } from './utils.js'
import getEnrichedInstanceData from '../methods/instanceData.js'
import type { BeforeScreenshotOptions, BeforeScreenshotResult } from './beforeScreenshot.interfaces.js'
import hideRemoveElements from '../clientSideScripts/hideRemoveElements.js'
import toggleTextTransparency from '../clientSideScripts/toggleTextTransparency.js'
import waitForFonts from '../clientSideScripts/waitForFonts.js'

const log = logger('@wdio/visual-service:beforeScreenshot')

/**
 * Methods that need to be executed before a screenshot will be taken
 */
export default async function beforeScreenshot(
    browserInstance: WebdriverIO.Browser,
    options: BeforeScreenshotOptions,
    addShadowPadding = false,
): Promise<BeforeScreenshotResult> {
    const { browserName, nativeWebScreenshot } = options.instanceData
    const { isAndroid, isIOS, isMobile } = browserInstance
    const {
        addressBarShadowPadding,
        disableBlinkingCursor,
        disableCSSAnimation,
        enableLayoutTesting,
        hideElements,
        noScrollBars,
        removeElements,
        toolBarShadowPadding,
        waitForFontsLoaded,
    } = options
    const addressBarPadding = getAddressBarShadowPadding({
        browserName,
        isAndroid,
        isIOS,
        isMobile,
        nativeWebScreenshot,
        addressBarShadowPadding,
        addShadowPadding,
    })
    const toolBarPadding = getToolBarShadowPadding({ isAndroid, isIOS, isMobile, browserName, toolBarShadowPadding, addShadowPadding })

    // Wait for the fonts to be loaded
    if (waitForFontsLoaded){
        try {
            await browserInstance.execute(waitForFonts)
        } catch (e) {
            log.debug('Waiting for fonts to load threw an error:', e)
        }
    }

    // Hide the scrollbars
    if (noScrollBars) {
        await browserInstance.execute(hideScrollBars, noScrollBars)
    }

    // Hide and or Remove elements
    if (hideElements.length > 0 || removeElements.length > 0) {
        try {
            await browserInstance.execute(hideRemoveElements, { hide: hideElements, remove: removeElements }, true)
        } catch (e) {
            log.warn(
                '\x1b[33m%s\x1b[0m',
                `
#####################################################################################
 WARNING:
 (One of) the elements that needed to be hidden or removed could not be found on the
 page and caused this error
 Error: ${e}
 We made sure the test didn't break.
#####################################################################################
`,
            )
        }
    }

    // Set some custom css
    if (disableCSSAnimation || disableBlinkingCursor || browserInstance.isMobile) {
        await browserInstance.execute(setCustomCss, { addressBarPadding, disableBlinkingCursor, disableCSSAnimation, id: CUSTOM_CSS_ID, toolBarPadding })
        // Wait at least 500 milliseconds to make sure the css is applied
        // Not every device is fast enough to apply the css faster
        await waitFor(500)
    }

    // Make all text transparent
    if (enableLayoutTesting){
        await browserInstance.execute(toggleTextTransparency, enableLayoutTesting)
        // Wait at least 500 milliseconds to make sure the css is applied
        // Not every device is fast enough to apply the css faster
        await waitFor(500)
    }

    // Get all the needed instance data
    const instanceOptions = {
        addressBarShadowPadding: options.addressBarShadowPadding,
        toolBarShadowPadding: options.toolBarShadowPadding,
        ...options.instanceData,
    }

    return getEnrichedInstanceData(browserInstance, instanceOptions, addShadowPadding)
}
