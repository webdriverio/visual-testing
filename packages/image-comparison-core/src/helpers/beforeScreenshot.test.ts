import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'
import { join } from 'node:path'
import logger from '@wdio/logger'
import beforeScreenshot from './beforeScreenshot.js'
import hideScrollBars from '../clientSideScripts/hideScrollbars.js'
import hideRemoveElements from '../clientSideScripts/hideRemoveElements.js'
import setCustomCss from '../clientSideScripts/setCustomCss.js'
import toggleTextTransparency from '../clientSideScripts/toggleTextTransparency.js'
import waitForFonts from '../clientSideScripts/waitForFonts.js'
import { CUSTOM_CSS_ID } from './constants.js'
import type { BeforeScreenshotOptions } from './beforeScreenshot.interfaces.js'

const log = logger('test')

vi.mock('@wdio/logger', () => import(join(process.cwd(), '__mocks__', '@wdio/logger')))

vi.mock('../methods/instanceData.js', () => ({
    default: vi.fn().mockResolvedValue({
        appName: 'mocked-app',
        browserName: 'mocked-browser',
        devicePixelRatio: 1,
        isAndroid: false,
        isIOS: false,
        isMobile: false,
        platformName: 'mocked-platform',
    })
}))

describe('beforeScreenshot', () => {
    let logDebugSpy: ReturnType<typeof vi.spyOn>
    let logWarnSpy: ReturnType<typeof vi.spyOn>

    const createMockBrowserInstance = (
        mockExecuteFn = vi.fn().mockResolvedValue(''),
        customProperties: Partial<WebdriverIO.Browser> = {}
    ) => {
        return {
            execute: mockExecuteFn,
            ...customProperties
        } as unknown as WebdriverIO.Browser
    }

    beforeEach(() => {
        logDebugSpy = vi.spyOn(log, 'debug').mockImplementation(() => {})
        logWarnSpy = vi.spyOn(log, 'warn').mockImplementation(() => {})
    })

    afterEach(() => {
        vi.clearAllMocks()
        logDebugSpy.mockRestore()
        logWarnSpy.mockRestore()
    })

    const baseInstanceData = {
        appName: 'appName',
        browserName: 'browserName',
        browserVersion: 'browserVersion',
        deviceName: 'deviceName',
        devicePixelRatio: 1,
        logName: 'logName',
        deviceRectangles: {
            bottomBar: { y: 0, x: 0, width: 0, height: 0 },
            homeBar: { x: 0, y: 0, width: 0, height: 0 },
            leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            screenSize: { height: 1, width: 1 },
            statusBar: { x: 0, y: 0, width: 0, height: 0 },
            statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
            viewport: { y: 0, x: 0, width: 0, height: 0 },
        },
        isAndroid: false,
        isIOS: false,
        isMobile: false,
        name: 'name',
        nativeWebScreenshot: false,
        platformName: 'platformName',
        platformVersion: 'platformVersion',
        initialDevicePixelRatio: 1,
    }
    const baseOptions = {
        addressBarShadowPadding: 6,
        disableBlinkingCursor: false,
        disableCSSAnimation: false,
        enableLayoutTesting: false,
        noScrollBars: false,
        toolBarShadowPadding: 6,
        hideElements: [],
        removeElements: [],
        waitForFontsLoaded: false,
    }
    const createOptions = (overrides: Partial<BeforeScreenshotOptions> = {}): BeforeScreenshotOptions => ({
        instanceData: baseInstanceData,
        ...baseOptions,
        ...overrides,
    })

    it('should be able to return the enriched instance data with default options', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const options = createOptions()

        expect(await beforeScreenshot(mockBrowserInstance, options)).toMatchSnapshot()
    })

    it('should be able to return the enriched instance data with `addShadowPadding: true`', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const options = createOptions({
            disableBlinkingCursor: true,
            disableCSSAnimation: true,
            noScrollBars: true,
            hideElements: [<HTMLElement>(<any>'<div></div>')],
            removeElements: [<HTMLElement>(<any>'<div></div>')],
            waitForFontsLoaded: true,
        })

        expect(await beforeScreenshot(mockBrowserInstance, options, true)).toMatchSnapshot()
    })

    it('should handle waitForFontsLoaded functionality', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const options = createOptions({
            waitForFontsLoaded: true,
        })

        await beforeScreenshot(mockBrowserInstance, options)

        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(waitForFonts)
    })

    it('should handle waitForFontsLoaded error gracefully and log debug message', async () => {
        const fontError = new Error('Font load error')
        const mockExecute = vi.fn().mockRejectedValue(fontError)
        const mockBrowserInstance = createMockBrowserInstance(mockExecute)
        const options = createOptions({
            waitForFontsLoaded: true,
        })

        await beforeScreenshot(mockBrowserInstance, options)

        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(waitForFonts)
        expect(logDebugSpy).toHaveBeenCalledWith('Waiting for fonts to load threw an error:', fontError)
    })

    it('should handle noScrollBars option', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const options = createOptions({
            noScrollBars: true,
        })

        await beforeScreenshot(mockBrowserInstance, options)

        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(hideScrollBars, true)
    })

    it('should handle hide and remove elements', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const hideElements = [<HTMLElement>(<any>'<div></div>')]
        const removeElements = [<HTMLElement>(<any>'<span></span>')]
        const options = createOptions({
            hideElements,
            removeElements,
        })

        await beforeScreenshot(mockBrowserInstance, options)

        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(hideRemoveElements, { hide: hideElements, remove: removeElements }, true)
    })

    it('should handle hide/remove elements error gracefully and log warning', async () => {
        const elementError = new Error('Element not found')
        const mockExecute = vi.fn().mockRejectedValue(elementError)
        const mockBrowserInstance = createMockBrowserInstance(mockExecute)
        const hideElements = [<HTMLElement>(<any>'<div></div>')]
        const options = createOptions({
            hideElements,
        })

        await beforeScreenshot(mockBrowserInstance, options)

        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(hideRemoveElements, { hide: hideElements, remove: [] }, true)
        expect(logWarnSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should handle CSS customization for desktop', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const options = createOptions({
            disableBlinkingCursor: true,
            disableCSSAnimation: true,
            addressBarShadowPadding: 10,
            toolBarShadowPadding: 15,
        })

        await beforeScreenshot(mockBrowserInstance, options)

        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(setCustomCss, {
            addressBarPadding: 0,
            disableBlinkingCursor: true,
            disableCSSAnimation: true,
            id: CUSTOM_CSS_ID,
            toolBarPadding: 0,
        })
    })

    it('should handle CSS customization for mobile platform', async () => {
        const mockBrowserInstance = createMockBrowserInstance(undefined, { isAndroid: true, isIOS: false, isMobile: true })
        const options = createOptions({
            instanceData: {
                ...baseInstanceData,
                platformName: 'Android',
                isAndroid: true,
                isMobile: true,
            },
        })

        await beforeScreenshot(mockBrowserInstance, options)

        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(setCustomCss, {
            addressBarPadding: 0,
            disableBlinkingCursor: false,
            disableCSSAnimation: false,
            id: CUSTOM_CSS_ID,
            toolBarPadding: 0,
        })
    })

    it('should handle layout testing', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const options = createOptions({
            enableLayoutTesting: true,
        })

        await beforeScreenshot(mockBrowserInstance, options)

        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(toggleTextTransparency, true)
    })

    it('should handle layout testing with addShadowPadding', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const options = createOptions({
            enableLayoutTesting: true,
            instanceData: {
                ...baseInstanceData,
                browserName: 'Safari',
                platformName: 'iOS',
                isIOS: true,
                isMobile: true,
            },
        })

        await beforeScreenshot(mockBrowserInstance, options, true)

        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(toggleTextTransparency, true)
    })

    it('should not execute browser commands when no options are enabled', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const options = createOptions()

        await beforeScreenshot(mockBrowserInstance, options)

        // Should only call the instanceData mock, no browser execute calls
        expect(mockBrowserInstance.execute).not.toHaveBeenCalled()
        expect(logDebugSpy).not.toHaveBeenCalled()
        expect(logWarnSpy).not.toHaveBeenCalled()
    })

    it('should handle multiple options simultaneously', async () => {
        const mockBrowserInstance = createMockBrowserInstance()
        const hideElements = [<HTMLElement>(<any>'<div></div>')]
        const options = createOptions({
            waitForFontsLoaded: true,
            noScrollBars: true,
            hideElements,
            disableBlinkingCursor: true,
            enableLayoutTesting: true,
        })

        await beforeScreenshot(mockBrowserInstance, options)

        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(waitForFonts)
        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(hideScrollBars, true)
        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(hideRemoveElements, { hide: hideElements, remove: [] }, true)
        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(setCustomCss, expect.any(Object))
        expect(mockBrowserInstance.execute).toHaveBeenCalledWith(toggleTextTransparency, true)
    })

    it('should handle multiple errors and log both debug and warning messages', async () => {
        const fontError = new Error('Font load error')
        const elementError = new Error('Element not found')
        const mockExecute = vi.fn()
            .mockRejectedValueOnce(fontError) // waitForFonts
            .mockRejectedValueOnce(elementError) // hideRemoveElements
        const mockBrowserInstance = createMockBrowserInstance(mockExecute)
        const options = createOptions({
            waitForFontsLoaded: true,
            hideElements: [<HTMLElement>(<any>'<div></div>')],
        })

        await beforeScreenshot(mockBrowserInstance, options)

        expect(logDebugSpy).toHaveBeenCalledWith('Waiting for fonts to load threw an error:', fontError)
        expect(logWarnSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should handle custom browser properties when needed', async () => {
        const mockBrowserInstance = createMockBrowserInstance(undefined, {
            getWindowSize: vi.fn().mockResolvedValue({ width: 1920, height: 1080 }),
            getOrientation: vi.fn().mockResolvedValue('LANDSCAPE')
        })
        const options = createOptions()

        await beforeScreenshot(mockBrowserInstance, options)

        expect(mockBrowserInstance.getWindowSize).toBeDefined()
        expect(mockBrowserInstance.getOrientation).toBeDefined()
    })
})
