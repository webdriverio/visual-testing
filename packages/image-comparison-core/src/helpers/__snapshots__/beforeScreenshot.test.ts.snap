// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`beforeScreenshot > should be able to return the enriched instance data with \`addShadowPadding: true\` 1`] = `
{
  "appName": "mocked-app",
  "browserName": "mocked-browser",
  "devicePixelRatio": 1,
  "isAndroid": false,
  "isIOS": false,
  "isMobile": false,
  "platformName": "mocked-platform",
}
`;

exports[`beforeScreenshot > should be able to return the enriched instance data with default options 1`] = `
{
  "appName": "mocked-app",
  "browserName": "mocked-browser",
  "devicePixelRatio": 1,
  "isAndroid": false,
  "isIOS": false,
  "isMobile": false,
  "platformName": "mocked-platform",
}
`;

exports[`beforeScreenshot > should handle hide/remove elements error gracefully and log warning 1`] = `
[
  "[33m%s[0m",
  "
#####################################################################################
 WARNING:
 (One of) the elements that needed to be hidden or removed could not be found on the
 page and caused this error
 Error: Error: Element not found
 We made sure the test didn't break.
#####################################################################################
",
]
`;

exports[`beforeScreenshot > should handle multiple errors and log both debug and warning messages 1`] = `
[
  "[33m%s[0m",
  "
#####################################################################################
 WARNING:
 (One of) the elements that needed to be hidden or removed could not be found on the
 page and caused this error
 Error: Error: Element not found
 We made sure the test didn't break.
#####################################################################################
",
]
`;
